pkg/ddevapp/router_test.go

```diff
@@ -266,3 +269,140 @@ func TestDisableHTTP2(t *testing.T) {
+	require.NoError(t, err)
+	defer l2.Close()
+
+	_, ok := ddevapp.FindAvailableRouterPort(startPort, badEndPort)
+	assert.Exactly(false, ok)
+
+	port, ok := ddevapp.FindAvailableRouterPort(startPort, goodEndPort)
+	assert.Exactly(true, ok)
+	assert.Exactly(startPort+3, port)
+}
+
+// Test that the app assigns an ephemeral port if the default one is not available.
+func TestUseEphemeralPort(t *testing.T) {
+	assert := asrt.New(t)
+
+	targetHTTPPort, targetHTTPSPort := "28080", "28443"
+
+	// This is copied from ddevapp_test.go
+	site1 := testcommon.TestSite{
+		Name:                          "TestEphemeralPort1", // Drupal D7
+		SourceURL:                     "https://ftp.drupal.org/files/projects/drupal-7.90.tar.gz",
+		ArchiveInternalExtractionPath: "drupal-7.90/",
+		FilesTarballURL:               "https://github.com/ddev/ddev_test_tarballs/releases/download/v1.1/d7test-7.59.files.tar.gz",
+		DBTarURL:                      "https://github.com/ddev/ddev_test_tarballs/releases/download/v1.1/d7test-7.87-db.tar.gz",
+		FullSiteTarballURL:            "",
+		Dir:                           testcommon.CreateTmpDir(t.Name() + "_1"),
+		Docroot:                       "",
+		Type:                          nodeps.AppTypeDrupal7,
+		Safe200URIWithExpectation:     testcommon.URIWithExpect{URI: "/README.txt", Expect: "Drupal is an open source content management platform"},
+		DynamicURI:                    testcommon.URIWithExpect{URI: "/node/1", Expect: "D7 test project, kittens edition"},
+		FilesImageURI:                 "/sites/default/files/field/image/kittens-large.jpg",
+		FullSiteArchiveExtPath:        "docroot/sites/default/files",
+	}
+
+	site2 := testcommon.TestSite{
+		Name:                          "TestEphemeralPort2",
+		SourceURL:                     "https://ftp.drupal.org/files/projects/drupal-8.9.20.tar.gz",
+		ArchiveInternalExtractionPath: "drupal-8.9.20/",
+		FilesTarballURL:               "https://github.com/ddev/ddev_test_tarballs/releases/download/v1.1/d8_umami.files.tar.gz",
+		FilesZipballURL:               "https://github.com/ddev/ddev_test_tarballs/releases/download/v1.1/d8_umami.files.zip",
+		DBTarURL:                      "https://github.com/ddev/ddev_test_tarballs/releases/download/v1.1/d8_umami.sql.tar.gz",
+		DBZipURL:                      "https://github.com/ddev/ddev_test_tarballs/releases/download/v1.1/d8_umami.sql.zip",
+		FullSiteTarballURL:            "",
+		Dir:                           testcommon.CreateTmpDir(t.Name() + "_2"),
+		Type:                          nodeps.AppTypeDrupal,
+		Docroot:                       "",
+		Safe200URIWithExpectation:     testcommon.URIWithExpect{URI: "/README.txt", Expect: "Drupal is an open source content management platform"},
+		DynamicURI:                    testcommon.URIWithExpect{URI: "/node/2", Expect: "Vegan chocolate and nut brownies"},
+		FilesImageURI:                 "/sites/default/files/vegan-chocolate-nut-brownies.jpg",
+	}
```

Comment:
There isn't actually a need for all this. We don't need (and aren't using all this TestSite stuff. All we need is a trivial tmp directory with a `ddev config` in it in order to test this.